Clean up current math parser

isuffix opened the pull request:
This is a reworking of the initial commits from #7072 and also includes the promised refactor for #7105.

This PR should not change any behavior of the math parser.

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -583,21 +575,21 @@ fn math_arg<'s>(p: &mut Parser<'s>, seen: &mut FxHashSet<&'s str>) -> bool {
     }
+    // We don't wrap when `count == 1`, since wrapping would change the type of
+    // the expression from potentially non-content to content.
+    // Ex: `$ func(#12pt) $` would change the type from size to content.
```
the newlines are a bit funky here. looks a bit like what happens with heavy rebase / jj usage? :p

isuffix replied:
This was meant to comment the end of the if-else chain, but since it didn't end in an `else` for `count == 1`, it looks odd.

Fixed by the other changes to spread args in the next comment.

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -526,52 +526,44 @@ fn math_arg<'s>(p: &mut Parser<'s>, seen: &mut FxHashSet<&'s str>) -> bool {
+            SyntaxNode::leaf(SyntaxKind::MathText, "."),
+            SyntaxNode::leaf(SyntaxKind::MathText, "."),
+        ];
+        p[m] = SyntaxNode::inner(SyntaxKind::Math, dots);
```
maybe not worth it, but just a thought: if we just push the text nodes flatly, we could maybe unify the `p.wrap(arg, SyntaxKind::Math);` for all branches?

laurmaedje reviewed (commented):
The changes look good! But there still seems to be a slight change in behavior in the last commit. Specifically, the output of the test case `$ f_1_2^3 $` which I mentioned [in my other review comment on the other PR](https://github.com/typst/typst/pull/7072#issuecomment-3385800159) still differs before & after the last commit.

I noticed a small change in how you handle the stop set compared to before, but unless I'm missing something, this didn't fix it.

isuffix commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -301,10 +291,10 @@ fn math_expr_prec(p: &mut Parser, min_prec: usize, stop: SyntaxKind) {
                 p.eat();
                 let m2 = p.marker();
-                math_expr_prec(p, 2, stop);
+                math_expr_prec(p, 2, syntax_set!());
```
Actually, because roots are a prefix operator, they shouldn't check against the caller's `min_prec` in the first place. Not sure why I didn't realize that before.

And it should use `MATH_ROOT_PREC`, good catch.

isuffix commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -253,7 +254,7 @@ fn math_expr_prec(p: &mut Parser, min_prec: usize, stop: SyntaxKind) {
             p.eat();
             // Parse a function call for an identifier or field access.
-            if min_prec < 3 && p.directly_at(SyntaxKind::LeftParen) {
+            if min_prec <= MATH_FUNC_PREC && p.directly_at(SyntaxKind::LeftParen) {
```
Yep. Setting to 2 :+1: 

Also, I guess you deleted your "the previous, flipped condition `prec >= min_prec` seemed clearer to me." comment? I still changed to using `prec >= min_prec` everywhere anyway because I think I agree that it's clearer. It also better matches the function comment: "Parses a math expression with at least the given precedence."

isuffix commented:
Rebased off main with the changes in the review comments. I also reordered the commits so that "Add new tests" comes first.

I definitely did not remember your comment on `$f_1_2^3$` from before :smiling_face_with_tear: . I've now added more variants of that case as the `math-attach-double-chain` test. This was tricky to get right because a naive solution would be to remove the `stop_set` check from the while loop condition (`while p.at_set(chain_set) && !p.at_set(stop_set)`), but removing that check breaks cases like `a^b'_c` in `math-primes-complex`. The actual fix is to only check the `stop_set` when _starting_ a chain with a prime, so that primes only start a chain when they don't _interrupt_ another chain.

The chaining loop now looks like this:
```rust
// Avoid interrupting a chain when initially parsing a prime.
// For `a^b'_c^d` the grouping is `(a^(b')_c)^d` and not `a^(b'_c^d)`.
if !(op_kind == SyntaxKind::MathPrimes && p.at_set(stop_set)) {
    // Parse chained attachment operators as a single attachment.
    while p.at_set(chain_set) {
        ...
    }
}
```

laurmaedje commented:
My first impression was that the change from 

https://github.com/typst/typst/blob/c85a6cdc3b3525d10d922853c202d942115a44f2/crates/typst-syntax/src/parser.rs#L357

to 

https://github.com/typst/typst/blob/593353e2666733bcb4c45a9986e8f7c5935d5da8/crates/typst-syntax/src/parser.rs#L363

(and same a bit below) was supposed to fix this issue. If it wasn't, what was that change about?

laurmaedje commented:
Thank you!
